### PR TITLE
⚡Backend Performance improvements

### LINKED ISSUE #146: Netlify builds are failing
Seen in #140 #145

@plibither8 @vikaspotluri123 @swapagarwal 

I think cherry-picking this commit would fix NPM failing to install packages. https://github.com/swapagarwal/swag-for-dev/pull/93/commits/26e36330f63fc1f1eb4b74432b43d5b3333ed49c

Gona have a try.

### PR DESCRIPTION
**Pull request by vikaspotluri123**
- Refs #92 (hopefully will close by the time I'm done)
- Todos are in the issue (I'm managing the issue so it's not too big of a deal)
- This PR is for backend improvements, I want to do frontend improvements separately to keep changes minimal

Closes #146
Closes #147

### CODE REVIEW

**Comment by aslafy-z** [comment]
The gulp pipeline is way cleaner that way ! Thank you @vikaspotluri123 :tada:

**Comment by vikaspotluri123** [comment]
todo: get rid of preload headers

File: site/get-data.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+const swagImages = [];
+const fileNames = [];
+const swagList = require('../data.json').map(swag => {
+    // Generate unique filename
+    // @todo: do we want to force jpeg as the image format?
```
Comment (aslafy-z) [inline site/get-data.js]: I forced jpeg in the initial code to standardize all images, maybe we could in the future create variants like webp to optimize their sizes even more but I'm not sure about keeping the original format if it's what you're asking about

File: site/get-data.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    // const extension = swag.image.split('.').pop();
```
Comment (aslafy-z) [inline site/get-data.js]: Also, this kind of cut is not really safe, it should rely on mimetype instead

File: site/get-data.js (hunk 1 above)
Comment (vikaspotluri123) [reply site/get-data.js]: I wasn't, but I was also thinking about that 😄 

I was asking more along the lines of if the image comes from the source as a PNG, is there a good reason to convert it to a jpeg (with the implication file type was actually considered by the source)?

File: site/get-data.js (hunk 2 above)
Comment (vikaspotluri123) [reply site/get-data.js]: True, but from what I can tell that adds quite a bit of complexity to the build system. While there could be a CI check (i.e. make sure the mime types match the extension) the complexity cost isn't worth it in the gulpfile

Comment (aslafy-z) [reply site/get-data.js]: I agree, that's why jpeg was forced

Comment (vikaspotluri123) [reply site/get-data.js]: We could also manually specify the format in the data file and default to jpeg if it's not provided 🤔

File: site/gulpfile.js [hunk 3]
```diff
@@ -1,137 +1,163 @@
...
-                }]
-            ]
-        }))
+        .pipe(babel({ presets }))
         .pipe(uglify())
         .pipe(gulp.dest('dist/assets/js'));
 });
 
+gulp.task('swag-img:clean', () => {
+    return del('dist/assets/swag-img/**/*');
```
Comment (aslafy-z) [inline site/gulpfile.js]: This path could lose two :star:s

File: site/gulpfile.js [hunk 4]
```diff
@@ -1,137 +1,163 @@
...
-            },
-        }, {
-            quality: 90,
-            progressive: true,
-            compressionLevel: 9,
-            errorOnEnlargement: false,
-            errorOnUnusedConfig: false,
-        }))
+        .pipe(responsive([{
+            name: '**/*',
```
Comment (aslafy-z) [inline site/gulpfile.js]: Two more :star:s to bust

File: site/gulpfile.js [hunk 5]
```diff
@@ -1,137 +1,163 @@
...
+gulp.task('webserver', () => {
+    return gulp.src('dist')
+        .pipe(webserver({
+            livereload: true,
+            open: true
+        }));
+});
 
 gulp.task('watch', () => {
     gulp.watch('src/pug/**/*.pug', gulp.parallel('pug'));
```
Comment (aslafy-z) [inline site/gulpfile.js]: Here again, :star:s

Sorry for the flood :fearful:

Comment (vikaspotluri123) [reply site/gulpfile.js]: This would need to be recurisve, but not js; if you change includes/body, pug needs to rebuild

**Comment by vikaspotluri123** [comment]
image
Did someone say caching 😎

File: site/gulpfile.js [hunk 6]
```diff
@@ -1,137 +1,163 @@
...
-    const tags = Array.from(builtSwagList.reduce(
+    const tags = Array.from(swagList.reduce(
         (tagList, { tags }) => {
             tags.forEach(tag => tagList.add(tag));
             return tagList;
         },
         new Set()
     ));
 
+    const manifest = require('./rev-manifest');
```
Comment (aslafy-z) [inline site/gulpfile.js]:
```suggestion
    const manifest = require('./rev-manifest.json');
```

as below.

File: site/gulpfile.js [hunk 7]
```diff
@@ -1,137 +1,163 @@
...
+        'dist/assets/swag-img/*'
+    ]);
 });
 
-gulp.task('swag-img', gulp.parallel('swag-img:build-data', gulp.series('swag-img:clean', 'swag-img:download' , 'swag-img:optimize')));
+gulp.task('cachebust', cb => {
+    const bustedFiles = [
+        'dist/assets/css/*',
+        'dist/assets/js/*',
+        'dist/assets/swag-img/*'
```
Comment (aslafy-z) [inline site/gulpfile.js]: This needs to be done for `dist/assets/img` also, right ?

Comment (aslafy-z) [reply site/gulpfile.js]: ok lgtm till we have more (if we have more)

**Comment by aslafy-z** [comment]
This is ready to deploy right? @vikaspotluri123

**Comment by vikaspotluri123** [comment]
Now there should be no blocker

**Comment by aslafy-z** [comment]
Actually, I think having `gulp-cli` installed locally is ok. It doesn't require users to install `gulp` globally before coding. `npm run build` is made possible because `gulp-cli` is installed.

**Comment by vikaspotluri123** [comment]
`npm run build` should use the module that ships with gulp 🤔 	

Either way, if we want to do that, I think it should be done as a separate PR since that's outside the scope of this one

**Comment by aslafy-z** [comment]
Actually you deleted the gulp-cli dependency but haven't added it. So removing the dep is outside of the scope of this PR.

Gulp does not provide any cli within the gulp package, AFAIK. Correct me if I'm wrong

**Comment by vikaspotluri123** [comment]
You're right 😳 

Looking at Gulp's deps: https://github.com/gulpjs/gulp/blob/v4.0.0/package.json#L34

No need to worry 😄 Since we're not using the CLI explicitly, we don't need it as a dependency 👍

**Comment by aslafy-z** [comment]
I was wrong! I thought we had to install it anyway. Good thing we don't have to! :+1: 

So, no more blocker right?

**Comment by vikaspotluri123** [comment]
From my side, we're good 👍

**Comment by aslafy-z** [comment]
Nice! Could you rebase and merge?

**Comment by vikaspotluri123** [comment]
I can rebase and merge, but haven't we been doing squashes?

**Comment by plibither8** [comment]
Yup, squash and merge.

**Comment by vikaspotluri123** [comment]
:shipit:

**Pull request merged**